docs-src/java/how-to-use-a-custom-payload-codec-in-java.md

```diff
@@ -0,0 +1,111 @@
+---
+id: how-to-use-a-custom-payload-codec-in-java
+title: How to use a custom Payload Codec in Java
+sidebar_label: Custom Payload Codec
+description: Create a custom implementation of `PayloadCodec` and use it in `CodecDataConverter` to set a custom Data Converter.
+tags:
+  - Java
+  - developer-guide
+---
+
+**Create a custom Payload Codec**
+
+Create a custom implementation of [`PayloadCodec`](https://www.javadoc.io/doc/io.temporal/temporal-sdk/latest/io/temporal/payload/codec/PayloadCodec.html) and use it in [`CodecDataConverter`](https://www.javadoc.io/doc/io.temporal/temporal-sdk/latest/io/temporal/common/converter/CodecDataConverter.html) to set a custom Data Converter.
+
+The Payload Codec does byte-to-byte conversion and must be set with a Data Converter.
+
+Define custom encryption/compression logic in your `encode` method and decryption/decompression logic in your `decode` method.
+
+The following example from the [Java encryption sample](https://github.com/temporalio/samples-java/blob/main/src/main/java/io/temporal/samples/encryptedpayloads/CryptCodec.java) shows how to implement encryption and decryption logic on your payloads in your `encode` and `decode` methods.
+
+```java
+class YourCustomPayloadCodec implements PayloadCodec {
+  //...
+
+  // See the linked sample for details on the methods called here.
+  @NotNull
+  @Override
+  public List<Payload> encode(@NotNull List<Payload> payloads) {
+    return payloads.stream().map(this::encodePayload).collect(Collectors.toList());
+  }
+
+  @NotNull
+  @Override
+  public List<Payload> decode(@NotNull List<Payload> payloads) {
+    return payloads.stream().map(this::decodePayload).collect(Collectors.toList());
+  }
+
+  private Payload encodePayload(Payload payload) {
+    String keyId = getKeyId(); // Get a fixed key Id from the getKeyId method. See the sample for details.
+    SecretKey key = getKey(keyId); // Get an AES key defined in the getKey method. See sample for details.
+
+    byte[] encryptedData;
+    try {
+      encryptedData = encrypt(payload.toByteArray(), key);
+    } catch (Throwable e) {
+      throw new DataConverterException(e);
+    }
+    // Apply metadata to the encoded payload that you can verify in your decode method before decoding.
+    // See the sample for details on the metadata values set.
+    return Payload.newBuilder()
+        .putMetadata(EncodingKeys.METADATA_ENCODING_KEY, METADATA_ENCODING)
```

Comment:
This piece of code is copy-pasted, the costants are still used, but why the constants themselves are not transferred together?

Reply:
Added the constants at the top, like in the sample. It looks cleaner.

Reply:
and added a comment inline for context.